File: pyrec/core/retrieval/content_based.cc
/*
 * Copyright 2020 cxsmarkchan. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "pyrec/core/retrieval/content_based.h"
#include "pyrec/proto/indexer.pb.h"

#include <boost/format.hpp>

namespace pyrec {
namespace service {

bool CBRetrievalServer::FillInvertedRequest(
    const pyrec::feature::FeatureList& feature_list,
    InvertedIndexerRequest* indexer_request) {
  if (feature_list.kind_case() != pyrec::feature::FeatureList::kBytesList)
    return false;

  for (auto& feature : feature_list.bytes_list().values()) {
    auto* search_request = indexer_request->add_search_requests();
    auto* search_items = search_request->mutable_search_items();
    (*search_items)[param_.extract_key.second] \
        .mutable_match_search_item() \
        ->set_bytes_value(feature);

    for (auto& filter_item : param_.filter_rules) {
      (*search_items)[filter_item.first] \
          .mutable_match_search_item() \
          ->set_bytes_value(filter_item.second);
    }
  }
  return true;
}

void CBRetrievalServer::FillReplyItems(const IndexerReply& indexer_reply,
                                       ItemReply* item_reply) {
  size_t num_items = indexer_reply.items_size();
  for (auto& index_item : indexer_reply.items()) {
    auto& field_map = index_item.fields().map_items();

    auto* reply_item = item_reply->add_items();
    auto* item_id_map = reply_item->mutable_item_id()->mutable_map_items();
    for (auto& key : param_.item_keys) {
      auto it_key = field_map.find(key);
      if (it_key == field_map.end())
        continue;

      (*item_id_map)[key].CopyFrom(it_key->second);
    }

    if (item_id_map->size() == 0) {
      item_reply->mutable_items()->RemoveLast();
      continue;
    }

    auto* retrieval_info = reply_item->add_retrieval_infos();
    retrieval_info->set_retrieval_id(retrieval_id_);
    retrieval_info->set_num_items(num_items);
  }
}

pyrec::Status CBRetrievalServer::RetrievalProcess(const PyRecRequest* request,
                                                  ItemReply* reply) {
  auto& context = request->context();
  auto it_context = context.find(param_.extract_key.first);
  if (it_context == context.end())
    return pyrec::Status::CANCELLED;
  auto& feature_map = it_context->second.map_items();

  auto it_feature = feature_map.find(param_.extract_key.second);
  if (it_feature == feature_map.end())
    return pyrec::Status::CANCELLED;
  auto& feature_list = it_feature->second;

  InvertedIndexerRequest indexer_request;
  indexer_request.set_max_num(request_num_);
  for (auto& key_id : param_.item_keys)
    indexer_request.add_requested_fields(key_id);
  FillInvertedRequest(feature_list, &indexer_request);
  IndexerReply indexer_reply;

  param_.indexer->CallInverted(indexer_request, &indexer_reply);

  reply->set_request_id(request->request_id());
  FillReplyItems(indexer_reply, reply);

  return pyrec::Status::OK;
}

}  // namespace service
}  // namespace pyrec

File: pyrec/retrieval/content_based.py
# Copyright 2020 cxsmarkchan. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Interfaces of content-based retrieval server."""
from pyrec.pywrap_core import CBRetrievalServerInterface

from pyrec.retrieval.retrieval_base import RetrievalServicePyBase
from pyrec.indexer.indexer_client import make_internal_indexer_client

class CBRetrievalServer(RetrievalServicePyBase):
  """Server of the retrieval service system."""
  def __init__(self):
    """Init function."""
    super().__init__()
    self._server = None
    self._retrieval_id = 0
    self._request_num = 0
    self._filter_rule = dict()
    self._extract_key = None
    self._indexer = None

  def set_retrieval_id(self, retrieval_id):
    """
    The retrieval id is used to identify the retrieval algorithm.
    :param retrieval_id:
    :return:
    """
    assert isinstance(retrieval_id, int)
    self._retrieval_id = retrieval_id
    return self

  def set_request_num(self, request_num):
    """
    The maximal number of items returned.
    :param request_num:
    :return:
    """
    assert isinstance(request_num, int)
    self._request_num = request_num
    return self

  def set_item_keys(self, item_keys):
    """
    Set the filter rules, which are used to retrieve only part of the items
    from the index.
    :param filter_rule:
    :return:
    """
    if isinstance(item_keys, int):
      self._item_keys = [item_keys]
    elif isinstance(item_keys, list):
      self._item_keys = item_keys
    else:
      raise TypeError(item_keys)

    return self


  def set_filter_rule(self, filter_rule):
    """
    Set the filter rules, which are used to retrieve only part of the items
    from the index.
    :param filter_rule:
    :return:
    """
    assert isinstance(filter_rule, dict)
    self._filter_rule = filter_rule
    return self

  def set_extract_key(self, extract_key):
    """
    Set the key used to retrieve items from the inverted index.
    :param extract_key:
    :return:
    """
    assert isinstance(extract_key, list)
    self._extract_key = extract_key
    return self

  def set_indexer(self, indexer):
    """
    Set the indexer containing all the items.
    """
    self._indexer = make_internal_indexer_client(indexer)
    return self

  def _create_server_impl(self):
    """
    Create the content-based retrieval server
    :return:
    """
    assert self._indexer is not None
    assert self._extract_key is not None

    self._server = CBRetrievalServerInterface()
    self._server.SetRetrievalId(self._retrieval_id) \
        .SetRequestNum(self._request_num) \
        .SetExtractKey(self._extract_key[0], self._extract_key[1]) \
        .SetIndexer(self._indexer)

    for key, value in self._filter_rule.items():
      self._server.AddFilterRule(key, value)

    for key in self._item_keys:
      self._server.AddItemKey(key)

    return self._server.Create() == 0

  def get_server(self):
    """
    Only used for transporting the server to a local client.
    :return: a reference of the server.
    """
    return self._server
